File: frontend/src/services/account.service.ts
import { apiClient } from "@/lib/axios";
import type { PagedResponse } from "@/types";
import type {
  Account,
  AccountFilterParams,
  AccountMovement,
  CreateAccountRequest,
} from "@/types/account.types";

const BASE = "/v1/accounts";

export const accountService = {
  findAll: async (params: AccountFilterParams): Promise<PagedResponse<Account>> => {
    const { data } = await apiClient.get<PagedResponse<Account>>(BASE, { params });
    return data;
  },

  findById: async (id: string): Promise<Account> => {
    const { data } = await apiClient.get<Account>(`${BASE}/${id}`);
    return data;
  },

  findByCustomerId: async (customerId: string): Promise<Account[]> => {
    const { data } = await apiClient.get<Account[]>(`${BASE}/customer/${customerId}`);
    return data;
  },

  create: async (payload: CreateAccountRequest): Promise<Account> => {
    const { data } = await apiClient.post<Account>(BASE, payload);
    return data;
  },

  getMovements: async (
    accountId: string,
    params?: { page?: number; size?: number },
  ): Promise<PagedResponse<AccountMovement>> => {
    const { data } = await apiClient.get<PagedResponse<AccountMovement>>(
      `${BASE}/${accountId}/movements`,
      { params },
    );
    return data;
  },

  freeze: async (id: string): Promise<Account> => {
    const { data } = await apiClient.patch<Account>(`${BASE}/${id}/freeze`);
    return data;
  },

  unfreeze: async (id: string): Promise<Account> => {
    const { data } = await apiClient.patch<Account>(`${BASE}/${id}/unfreeze`);
    return data;
  },

  close: async (id: string): Promise<Account> => {
    const { data } = await apiClient.patch<Account>(`${BASE}/${id}/close`);
    return data;
  },
};

File: backend/src/main/java/com/corebanking/modules/account/service/AccountService.java
package com.corebanking.modules.account.service;

import com.corebanking.modules.account.dto.AccountFilterParams;
import com.corebanking.modules.account.dto.AccountMovementResponse;
import com.corebanking.modules.account.dto.AccountResponse;
import com.corebanking.modules.account.dto.CreateAccountRequest;
import com.corebanking.shared.PagedResponse;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

public interface AccountService {

    PagedResponse<AccountResponse> findAll(AccountFilterParams filters, Pageable pageable);

    AccountResponse findById(UUID id);

    List<AccountResponse> findByCustomerId(UUID customerId);

    AccountResponse create(CreateAccountRequest request);

    AccountResponse freeze(UUID id);

    AccountResponse unfreeze(UUID id);

    AccountResponse close(UUID id);

    PagedResponse<AccountMovementResponse> getMovements(UUID accountId, Pageable pageable);

    /**
     * Used by TransferService (Phase 5) within its own @Transactional boundary.
     * Validates active account, sufficient balance, and registers a RETIRO movement.
     */
    AccountResponse debit(UUID accountId, BigDecimal amount, String description, String reference);

    /**
     * Used by TransferService (Phase 5) within its own @Transactional boundary.
     * Validates active account and registers a DEPOSITO movement.
     */
    AccountResponse credit(UUID accountId, BigDecimal amount, String description, String reference);
}
